File: solutions_1-50/problem-2/even-fibonacci-numbers/src/main.cpp
#include <iostream>
using namespace std;

int fib(int);

int main()
{
    int sum = 0;
    int i = 0;
    int f = fib(i);
    while (f < 4000000)
    {
        cout << i << " " << f << '\n';
        if (f % 2 == 0)
            sum += f;
        i++;
        f = fib(i);
    }
    cout << sum << '\n';
    return 0;
}

int fib(int n)
{
    if (n <= 2)
        return (n % 2 + n) / 2; // 0->0, 1->1, 2->1
    return fib(n - 1) + fib(n - 2);
}

File: solutions_1-50/problem-2/even-fibonacci-numbers/src/main.rs
fn main() {
    let mut sum: u32 = 0;
    let mut i: u32 = 0;
    let mut f: u32 = fib(i);
    while f < 4000000 {
        println!("{} {}", i, f);
        if f.is_multiple_of(2) {
            sum += f;
        }
        i += 1;
        f = fib(i);
    }
    println!("{}", sum);
}

fn fib(n: u32) -> u32 {
    if n <= 2 {
        return (n % 2 + n) / 2; // 0->0, 1->1, 2->1
    }
    fib(n - 1) + fib(n - 2)
}
